### PR TITLE
MP Connect dialog overhaul and server management UI for #4564

### LINKED ISSUE #4564: Add a UI to manage custom servers
Currently the only known way of adding a custom server to the MP dialog ("View List" button under "Connect to Server") is to manually edit the game_config.cfg. It would be much nicer if there was a UI to manage this which stores the result in the preferences.

### PR DESCRIPTION
**Pull request by shikadiqueen**
There has been functionality to manage custom servers in the player's preferences file since as far back as commit 5fb0b2a91123474d147c9de4918fc11a1b4c15ac in June 2007, but there's never been a user interface for it.

image

This commit removes the current server list dialog attached to the MP Connect dialog and replaces it with the same functionality embedded into the latter, as well as the ability to edit the server list in the player's preferences file. There's some code movement in a few other places since the previous function in the preferences namespace doesn't suit the new code (it always returned a combined list of game and player-defined servers).

There's some disabled code referencing a situation where the listbox with the list of servers allows the code (and the player) to clear the selection. It's disabled because there are some UX issues with it noted in the code comments associated to it.

Fixes #4564.

### CODE REVIEW

File: src/gui/dialogs/edit_text.cpp [hunk 1]
```diff
@@ -41,14 +46,37 @@ namespace dialogs
...
+   }
+}
+
+void edit_text::on_text_change()
+{
+   text_box& text = find_widget<text_box>(get_window(), "text", false);
+   button& ok_button = find_widget<button>(get_window(), "ok", false);
+
+   auto clean_value = text.get_value();
+	boost::trim(clean_value);
```
Comment (Vultraz) [inline src/gui/dialogs/edit_text.cpp]: You don't need to create a temporary here. You can use `boost::trim_copy`.

File: src/gui/dialogs/edit_text.cpp [hunk 2]
```diff
@@ -41,14 +46,37 @@ namespace dialogs
...
+   : disallow_empty_(disallow_empty)
 {
 	register_label("title", true, title, true);
 	register_label("label", true, label, true);
 	register_text("text", true, text, true);
 }
+
+void edit_text::pre_show(window& window)
+{
+   if(disallow_empty_) {
```
Comment (Vultraz) [inline src/gui/dialogs/edit_text.cpp]: Bunch of spaces indent around here.

File: src/gui/dialogs/multiplayer/mp_connect.cpp [hunk 3]
```diff
@@ -16,174 +16,304 @@
...
+void clear_listbox_selection(listbox& listbox)
+{
+	const auto selection = listbox.get_selected_row();
+	if(selection >= 0) {
+		listbox.select_row(selection, false);
+	}
+}
+#endif
+
+std::string clean_textbox_value(text_box& text_box)
```
Comment (Vultraz) [inline src/gui/dialogs/multiplayer/mp_connect.cpp]: You don't need this, you can use `boost::trim_copy`.

File: src/gui/dialogs/multiplayer/mp_connect.cpp [hunk 4]
```diff
@@ -16,174 +16,304 @@
...
+		{ "address", string_map{{"label", srv.address}} },
+	};
+
+	listbox.add_row(entry, pos);
+}
+
+void mp_connect::select_first_match()
+{
+	window* window = get_window();
+	if(!window) {
```
Comment (Vultraz) [inline src/gui/dialogs/multiplayer/mp_connect.cpp]: You don't need to always check `window` is valid. It's guaranteed to exist as long as the dialog is shown.

File: src/gui/dialogs/multiplayer/mp_connect.cpp [hunk 5]
```diff
@@ -16,174 +16,304 @@
...
-	listbox& list = find_widget<listbox>(&window, "server_list", false);
+	// Select the first matching list entry or clear the current selection
+	select_first_match();
+}
 
-	window.keyboard_capture(&list);
+void mp_connect::on_server_add()
+{
+	window* window = get_window();
+	if(!window) {
```
Comment (Vultraz) [inline src/gui/dialogs/multiplayer/mp_connect.cpp]: Ditto about checking `window`.

File: src/gui/dialogs/multiplayer/mp_connect.cpp [hunk 6]
```diff
@@ -16,174 +16,304 @@
...
+	insert_into_server_listbox(server_list, info, ui_pos);
+	select_first_match();
 }
 
-void mp_server_list::post_show(window& window)
+void mp_connect::on_server_delete()
 {
-	if(get_retval() == retval::OK) {
+	window* window = get_window();
+	if(!window) {
```
Comment (Vultraz) [inline src/gui/dialogs/multiplayer/mp_connect.cpp]: Ditto about checking `window`.

File: src/gui/dialogs/multiplayer/mp_connect.cpp [hunk 7]
```diff
@@ -16,174 +16,304 @@
...
+	on_server_select();
+}
 
-static void
-show_server_list(window& window, field_text* host_name)
+void mp_connect::on_server_select()
 {
-	assert(host_name);
+	window* window = get_window();
+	if(!window) {
```
Comment (Vultraz) [inline src/gui/dialogs/multiplayer/mp_connect.cpp]: Ditto about checking `window`.

File: src/preferences/game.cpp [hunk 8]
```diff
@@ -380,28 +380,43 @@ void _set_lobby_joins(int show)
...
+
+std::vector<game_config::server_info> user_servers_list()
+{
+	std::vector<game_config::server_info> pref_servers;
+
+	for(const config &server : get_prefs()->child_range("server")) {
+		game_config::server_info sinf;
+		sinf.name = server["name"].str();
+		sinf.address = server["address"].str();
+		pref_servers.push_back(sinf);
```
Comment (Vultraz) [inline src/preferences/game.cpp]: You can use `emplace_back` here.

File: src/gui/dialogs/multiplayer/mp_connect.cpp [hunk 9]
```diff
@@ -16,174 +16,304 @@
...
+#endif
+	button_add.set_active(!address.empty());
+}
 
-void mp_server_list::pre_show(window& window)
+void mp_connect::on_address_change()
 {
-	set_restore(true);
+	window* window = get_window();
+	if(!window) {
```
Comment (Vultraz) [inline src/gui/dialogs/multiplayer/mp_connect.cpp]: Ditto about checking window.

File: src/preferences/game.cpp (hunk 8 above)
Comment (shikadiqueen) [reply src/preferences/game.cpp]: There's a (redundant) default constructor explicitly defined for `game_config::server_info` which interferes with this and other things such as move assignment.

**Comment by shikadiqueen** [comment]
Addressed all issues except for the `emplace_back` recommendation since the class' custom default ctor gets in the way. I may append that to the series later today, though.

**Comment by shikadiqueen** [comment]
Sorry, that wasn't quite right. I forgot a `boost::trim` in the edit text dialog.

**Comment by shikadiqueen** [comment]
My mistake, the `emplace_back` issue is simply because the class doesn't define a constructor that takes two strings as arguments. Forcing use of the empty ctor and working on the result is a viable alternative though.

**Comment by CelticMinstrel** [comment]
Looks good, would be nice if we can solve the problems with allowing an empty selection in the list though. If I understand correctly, what happened was you type in a server that's not in the list, which then causes the selection to be cleared, which then means … what exactly? Clicking the plus button with an empty selection should probably append to the end, but it sounds like you want it to insert where the selection just was? In that case, maybe you could just keep the current selection if the newly-typed server doesn't match.

**Comment by shikadiqueen** [comment]
> If I understand correctly, what happened was you type in a server that's not in the list, which then causes the selection to be cleared, which then means … what exactly? Clicking the plus button with an empty selection should probably append to the end, but it sounds like you want it to insert where the selection just was? In that case, maybe you could just keep the current selection if the newly-typed server doesn't match.

Personally I believe that it makes sense to always show where the selection is at when it comes to these kind of things. If the selection is cleared then you're assuming that the player will remember where it was at when you perform an operation next while you keep track of it behind the scenes. For me, any situation where the user interface *knows* things that aren't being visibly displayed is an attack on the user.

Now, technically, adding right below the current/last selection isn't a strictly necessary feature. Appending always to the end of the list would work just as well, especially since this isn't going to see as much use as for people to wind up in a situation where having to rearrange the list by deleting and re-adding entries is going to become a concrete problem. We could even have up/down buttons to rearrange list items, stealing more real estate and making the UI seem cramped.

I'd like to think that the decision I made (there's always a selection even if it doesn't match) is a better compromise than the alternative from a UX point of view, since ultimately the authoritative source of the server address is always going to be whatever is displayed on the textbox.

**Comment by CelticMinstrel** [comment]
The only situation where I think I might disagree is when you first open the window and the remembered server doesn't match anything on the list.

**Comment by shikadiqueen** [comment]
That doesn't seem important to me.

On the other hand, if you feel inclined to write code to disallow the player form clearing the selection themselves by clicking on it that would be a reasonable compromise.

**Comment by CelticMinstrel** [comment]
Yeah… well, I'll think about it. If I feel like it, maybe I'll do that at some point after this is merged.

File: src/gui/dialogs/edit_text.cpp (hunk 1 above)
Comment (AI0867) [reply src/gui/dialogs/edit_text.cpp]: The temporary is still created, though it is no longer used.

**Pull request merged**